### PR TITLE
feat: finalise mls migration

### PR DESCRIPTION
**Pull request by PatrykBuniX**
Finalises the MLS migration of conversations that meet the requirements. It will finalise if either of the 2 is truthy
- migration was finalised - `finaliseRegardlessAfter` date has arrived, or
- all active clients of all the conversation participants support MLS (`supportedProtocols` fields on user entity includes `MLS`)

To finalise, we just change the protocol to `MLS` on both backend and local store.
From this point on, messages will be sent via MLS.

### CODE REVIEW

File: src/script/mls/MLSMigration/MLSMigration.ts [hunk 1]
```diff
@@ -137,40 +135,33 @@ const checkMigrationConfig = async (
...
   //TODO: implement logic for 1on1 conversations (both team owned and federated)
-  const regularGroupConversations = conversations.filter(
-    conversation =>
-      conversation.type() === CONVERSATION_TYPE.REGULAR &&
-      isConversationOwnedBySelfTeam(conversation) &&
-      !conversation.isTeam1to1(),
-  );
-
-  const {proteus: proteusConversations, mixed: mixedConversatons} =
-    groupConversationsByProtocol(regularGroupConversations);
```
Comment (PatrykBuniX) [inline src/script/mls/MLSMigration/MLSMigration.ts:165]: We cannot filter conversations by protocols here as the protocol might change during migration (proteus -> mixed -> mls). We're filtering conversations by protocol inside initialise, joinmixed and finalise methods instead.

File: src/script/mls/MLSMigration/initialiseMigration/tryEstablishingMLSGroupForMixedConversation/tryEstablishingMLSGroupForMixedConversation.ts [hunk 2]
```diff
@@ -54,22 +54,21 @@ export const tryEstablishingMLSGroupForMixedConversation = async (
...
-    client: core.clientId,
-  });
-
-  //if there's no response, it means that commit bundle was not sent successfully
-  //at this point we should wipe conversation locally
-  //it's possible that somebody else has already created the group,
-  //we should wait for the welcome message or try joining with external commit later
-  if (!groupCreationResponse) {
+  try {
+    await mlsService.registerConversation(groupId, [], {
```
Comment (PatrykBuniX) [inline src/script/mls/MLSMigration/initialiseMigration/tryEstablishingMLSGroupForMixedConversation/tryEstablishingMLSGroupForMixedConversation.ts:58]: This method does no longer return null if uploading commit bundle fails, instead an error is thrown, so we have to catch it.

File: src/script/util/DebugUtil.ts [hunk 3]
```diff
@@ -195,6 +198,26 @@ export class DebugUtil {
     await proteusService['cryptoClient'].debugBreakSession(sessionId);
   }
 
+  async setTeamSupportedProtocols(supportedProtocols: ConversationProtocol[]) {
```
Comment (PatrykBuniX) [inline src/script/util/DebugUtil.ts:201]: Will enable QA to start testing the feature before TM UI for `supportedProtocols` is ready.

File: src/script/mls/MLSMigration/finaliseMigration/finaliseMigration.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+const checkFinalisationCriteria = async (
+  mixedConversation: MixedConversation,
+  onReadyToFinalise: (mixedConversation: MixedConversation) => Promise<void>,
+  {teamState}: {teamState: TeamState},
+) => {
+  const mlsMigrationFeature = teamState.teamFeatures().mlsMigration;
+  const migrationStatus = getMLSMigrationStatus(mlsMigrationFeature);
+  const isMigrationFinalised = migrationStatus === MLSMigrationStatus.FINALISED;
+
+  if (isMigrationFinalised || doAllConversationParticipantsSupportMLS(mixedConversation)) {
```
Comment (PatrykBuniX) [inline src/script/mls/MLSMigration/finaliseMigration/finaliseMigration.ts:57]: According to https://wearezeta.atlassian.net/wiki/spaces/CORE/pages/759890302/Use+case+client+checks+if+conversation+migration+can+be+finalised+Proteus+to+MLS+migration we are ready to finalise the migration of conversation when all of the conversation participants support mls (add mls to the list of their supported protocols) or when migration is finalised (`finaliseRegardlessAfter` migration force date arrives)

File: src/script/mls/supportedProtocols/evaluateSelfSupportedProtocols/evaluateSelfSupportedProtocols.ts [hunk 5]
```diff
@@ -52,13 +52,21 @@ export const evaluateSelfSupportedProtocols = async ({
     supportedProtocols.add(ConversationProtocol.PROTEUS);
   }
 
-  if (await isMLSSupported({teamSupportedProtocols, selfClients, mlsMigrationStatus, core, apiClient})) {
+  if (
+    (await isMLSSupported({teamSupportedProtocols, selfClients, mlsMigrationStatus, core, apiClient})) ||
+    (await isMLSForcedWithoutMigration({teamSupportedProtocols, selfClients, mlsMigrationStatus, core, apiClient}))
```
Comment (przemvs) [inline src/script/mls/supportedProtocols/evaluateSelfSupportedProtocols/evaluateSelfSupportedProtocols.ts]: Move this to variables, if will be more readable :D

File: src/script/user/UserRepository.ts [hunk 6]
```diff
@@ -750,6 +750,15 @@ export class UserRepository {
     return users.map(user => this.updateSavedUser(user));
   }
 
+  /**
+   * Refresh all known users (in local state) from the backend.
+   */
+  async refreshAllKnownUsers() {
+    const userIds = this.userState.users().map(user => user.qualifiedId);
+    const {found: users} = await this.fetchRawUsers(userIds, this.userState.self().domain);
+    return users.map(user => this.updateSavedUser(user));
```
Comment (przemvs) [inline src/script/user/UserRepository.ts]: You need to map users there and return them? Can't use forEach for e.g ?

Comment (przemvs) [reply src/script/user/UserRepository.ts]: And I think You can extend this function using refreshUsers function :)

Comment (PatrykBuniX) [reply src/script/user/UserRepository.ts]: Good idea 😄

File: src/script/mls/MLSMigration/finaliseMigration/finaliseMigration.test.ts [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+const createMixedConversation = (): MixedConversation => {
+  const conversation = new Conversation(createUuid(), '', ConversationProtocol.MIXED);
+  const mockGroupId = 'groupId';
+  conversation.groupId = mockGroupId;
+  conversation.type(CONVERSATION_TYPE.REGULAR);
+  return conversation as MixedConversation;
+};
+
+const turnConversationToMLS = (conversation: MixedConversation): MLSConversation => {
```
Comment (przemvs) [inline src/script/mls/MLSMigration/finaliseMigration/finaliseMigration.test.ts]: I think replace will be better instead of turn :) WDYT?

Comment (przemvs) [reply src/script/mls/MLSMigration/finaliseMigration/finaliseMigration.test.ts]: Or change

File: src/script/mls/supportedProtocols/evaluateSelfSupportedProtocols/evaluateSelfSupportedProtocols.ts [hunk 8]
```diff
@@ -48,17 +48,39 @@ export const evaluateSelfSupportedProtocols = async ({
...
+    apiClient,
+  });
+
+  const isMLSForced = await isMLSForcedWithoutMigration({
+    teamSupportedProtocols,
+    selfClients,
+    mlsMigrationStatus,
+    core,
+    apiClient,
+  });
```
Comment (przemvs) [inline src/script/mls/supportedProtocols/evaluateSelfSupportedProtocols/evaluateSelfSupportedProtocols.ts]: Move also props to variable :D It's same props passed to this function :)

File: src/script/user/UserRepository.ts [hunk 9]
```diff
@@ -750,6 +750,14 @@ export class UserRepository {
     return users.map(user => this.updateSavedUser(user));
   }
 
+  /**
+   * Refresh all known users (in local state) from the backend.
+   */
+  async refreshAllKnownUsers() {
+    const userIds = this.userState.users().map(user => user.qualifiedId);
+    return this.refreshUsers(userIds);
```
Comment (przemvs) [inline src/script/user/UserRepository.ts:758]: ❤️

**Review by przemvs — APPROVED** [review]

**Pull request merged**